### ISSUE TITLE
reduce_mean is not numerically accurate on GPU

### ISSUE THREAD

**Issue by zhengwy888** [tensorflow/tensorflow#3775]
For smaller size matrix tf.reduce_mean works fine. 
But for larger size matrix floating point results are not accurate.

I am running on CUDA 7.0, Titan X.

```python
import tensorflow as tf

mat3 = tf.get_variable("mat3", [4500,4500], initializer=tf.constant_initializer(-1.))
reduce_mat3_ref = tf.reduce_mean(mat3)
with tf.Session('') as sess:
    tf.initialize_all_variables().run()
    reduce1 = sess.run(reduce_mat3_ref)
    print(reduce1) # on CPU this is -1.0, on GPU this produces -1.00006
    assert(reduce1 == -1.) 
```

**Comment by yaroslavvb** [comment]
Also on CUDA 8.0, GTX 980/GTX 1080, result is off by 5.86509705e-05

**Comment by yaroslavvb** [comment]
Relevant issue: #2625 
Although I don't think numpy uses any of those stable methods, and it still gives exact result on the problem above, so something else is going on

**Comment by sherrym** [comment]
There's a PR for Eigen to support double at the moment:

https://bitbucket.org/eigen/eigen/pull-requests/216/enable-efficient-tensor-reduction-for/diff

You can add yourself to the interest list of that PR. As soon as that's available, you can explicitly make your input tensor to be of dtype double and get the accuracy you prefer. How does that sound?

**Comment by zhengwy888** [comment]
the problem using both float on GPU and CPU, the results does not agree on each other. and this example is trivial enough, all the numbers are -1 in the entry. Float's precision should have been sufficient.

**Comment by tatatodd** [comment]
I believe this is a duplicate of #3103, which was closed as "working as intended".

The basic problem is that there's a speed/accuracy tradeoff for reduce_mean.  Even though your example seems trivial, with the entire matrix initialized to -1.0, there are 20M+ entries (4500*4500).  Note that float32 can represent integral numbers in the range [-16M,+16M] without loss of precison (24 bits):
https://en.wikipedia.org/wiki/Single-precision_floating-point_format

So we expect to see some error if we accumulate all of the -1.0 entries and divide at the end to compute the mean.  This also explains why smaller matrices didn't exhibit the problem.

Hope this answers your question!
